test/ossl_shim/packeted_bio.h

```diff
@@ -0,0 +1,49 @@
+/* Copyright (c) 2014, Google Inc.
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+ * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE. */
+
+#ifndef HEADER_PACKETED_BIO
+#define HEADER_PACKETED_BIO
+
+#include <openssl/base.h>
+#include <openssl/bio.h>
+
+#if defined(OPENSSL_WINDOWS)
+OPENSSL_MSVC_PRAGMA(warning(push, 3))
+#include <winsock2.h>
+OPENSSL_MSVC_PRAGMA(warning(pop))
+#else
+#include <sys/time.h>
+#endif
+
+
+// PacketedBioCreate creates a filter BIO which implements a reliable in-order
+// blocking datagram socket. It internally maintains a clock and honors
+// |BIO_CTRL_DGRAM_SET_NEXT_TIMEOUT| based on it.
```

Comment:
I think retransmit tests and being able to mock out the clock are a pretty reasonable things to want, so I'd maybe suggest leaving it there and adding the callback? WebRTC even asked us to expose it in public API because they found it useful for their own tests. The machinery was later useful in fuzzing for stripping the last of the non-determinism (timestamps in sessions) which is nice when recording transcripts[*].

That said, the retransmit tests are very particular to BoringSSL's retransmit strategy. I think they currently match, but those are probably a very early set to move from "external tests" to an OpenSSL-tailored suite, unless we come up with clever `shim_config.json` spellings.

If you don't end up using that, you can trim basically all of packeted BIO. At that point, packeted BIO is just the length-prefix thing so packet boundaries are preserved.

[*] See https://boringssl.googlesource.com/boringssl/+/HEAD/FUZZING.md#TLS-transcripts for our nascent fuzzing strategy which seeds from this test suite.